File: web3/contracts/profile_contract.py
from pyteal import *

def profile_smart_contract():
    # Define state keys
    fullname_key = Bytes("fullname")
    username_key = Bytes("username")
    gender_key = Bytes("gender")
    bio_key = Bytes("bio")

    # Define transactions
    @Subroutine(TealType.uint64)
    def create_update_profile(fullname: Expr, username: Expr, gender: Expr, bio: Expr):
        return Seq([
            App.localPut(Int(0), fullname_key, fullname),
            App.localPut(Int(0), username_key, username),
            App.localPut(Int(0), gender_key, gender),
            App.localPut(Int(0), bio_key, bio),
            Int(1)
        ])

    # Main program logic
    program = Cond(
        [Txn.application_id() == Int(0), Return(Int(1))],  # Creation
        [Txn.on_completion() == OnComplete.NoOp, Return(create_update_profile(
            Txn.application_args[0],  # fullname
            Txn.application_args[1],  # username
            Txn.application_args[2],  # gender
            Txn.application_args[3]   # bio
        ))],
        [Txn.on_completion() == OnComplete.OptIn, Return(Int(1))],  # Handle opt-in
        [Int(1), Return(Int(0))]  # Default case, return 0 for failure
    )

    return program

def clear_state_program():
    return Return(Int(1))

if __name__ == "__main__":
    with open("approval_program.teal", "w") as f:
        compiled_approval_program = compileTeal(profile_smart_contract(), mode=Mode.Application, version=4)
        f.write(compiled_approval_program)

    with open("clear_state_program.teal", "w") as f:
        compiled_clear_state_program = compileTeal(clear_state_program(), mode=Mode.Application, version=4)
        f.write(compiled_clear_state_program)

File: web3/contracts/profilecontract.js
const algosdk = require('algosdk');
const fs = require('fs');

// Algorand Testnet PureStake API Information
const algodToken = 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa';
const algodServer = 'http://localhost';
const algodPort = '4001';

// Your account mnemonics
const senderMnemonic = 'story keen gorilla slide question skill derive knock become venture winner omit faint spell disorder puzzle very rifle worry still combine write fitness absent afraid';

// Instantiate the algod wrapper
let algodClient = new algosdk.Algodv2(algodToken, algodServer, algodPort);

// Function to compile program source
async function compileProgram(client, programSource) {
  let compileResponse = await client.compile(programSource).do();
  let compiledBytes = new Uint8Array(Buffer.from(compileResponse.result, "base64"));
  return compiledBytes;
}

async function deployContract(client, sender, approvalProgram, clearStateProgram) {
  // Get network params
  const params = await client.getTransactionParams().do();

  const numGlobalByteSlices = 0;
  const numGlobalInts = 1;
  const numLocalByteSlices = 4; // For fullname, username, gender, bio
  const numLocalInts = 0;
  

  // Create an application creation transaction
  const txn = algosdk.makeApplicationCreateTxn(
    sender.addr,
    params,
    algosdk.OnApplicationComplete.NoOpOC,
    approvalProgram,
    clearStateProgram,
    numLocalInts,
    numLocalByteSlices,
    numGlobalInts,
    numGlobalByteSlices,
    [],
    []
  );

  // Sign the transaction
  const signedTxn = txn.signTxn(sender.sk);

  // Send the transaction
  const sendResponse = await client.sendRawTransaction(signedTxn).do();

  // Wait for confirmation
  const confirmedTxn = await algosdk.waitForConfirmation(client, sendResponse.txId, 4);

  // Get the application ID
  const appId = confirmedTxn["application-index"];
  return appId;
  
}


async function main() {
  // Read and compile TEAL programs
  const approvalProgramSource = fs.readFileSync('approval_program.teal', 'utf8');
  const compiledApprovalProgram = await compileProgram(algodClient, approvalProgramSource);
  const clearStateProgramSource = fs.readFileSync('clear_state_program.teal', 'utf8');
  const compiledClearStateProgram = await compileProgram(algodClient, clearStateProgramSource);

  // Recover the account from the mnemonic
  // const sender = algosdk.mnemonicToSecretKey(senderMnemonic);
  const sender = algosdk.mnemonicToSecretKey(senderMnemonic);

  // Deploy the smart contract
  const appId = await deployContract(algodClient, sender, compiledApprovalProgram, compiledClearStateProgram);
  console.log("Deployed Smart Contract with App ID:", appId);
}

main().catch(console.error);
